### ISSUE TITLE
Userdata Qos usecase

### ISSUE THREAD

**Issue by raiars** [eclipse-cyclonedds/cyclonedds#54]
Hello, I've been testing with userdata qos policy. The project currently has the ability to set the userdata qos. But how can a remote discovered participant get the qos policy I set before? Thanks!

**Comment by raiars** [comment]
Thank you so much for explanation!

**Comment by raiars** [comment]
One more question. When I am debugging, I find that the function forward_builtin_participant doesn't work. Step into it until write_impl function, it shows that the builtin writers for both CMParticipantBuiltinTopicData and ParticipantBuiltinTopicData topics have no matched readers to send, which means the builtin writers don't match with the readers. Is there any plan to fix this problem?

**Comment by eboasson** [comment]
If you stepped through the first invocation of forward_builtin_participant, then there shouldn't be any readers yet. You can only create a reader once you have a participant, and so the first call to forward_builtin_participant necessarily occurs when there are none yet. This is not a problem because the data is transient-local, and hence it will be stored in the writer's history cache. When the reader (which should also be transient-local) is created it gets the data from there.

Does that answer your question?

**Comment by jwcesign** [comment]
The participant discovery data of pub is sent Periodically, but I find the sub will just get the first  participant discovery data of pub, I guess there is a checking about whether it is a new participant and then handle the data, can u tell me how u make this?

**Comment by eboasson** [comment]
You're right: pretty much the first thing it does when it receives an SPDP message is to check whether the participant is already known. It does that in ``handle_SPDP_alive``, where it distinguishes four cases:

- a local participant (usually the multicasts are looped back, so you receive your own multicasts), in which case it ignores it; it is a local participant if ``ephash_lookup_participant_guid`` finds something;
- a known remote participant, in which case it renews the lease; this is if ``ephash_lookup_proxy_participant_guid`` finds something;
- a recently deleted participant (local or remote, so it can suppress rediscovery for some time), in which case it ignores it; this is checked by ``is_deleted_participant_guid``
- an entirely unknown participant, in which case it creates a new proxy participant.

In case you wonder why there are these two lookups instead of a single one: there used to be separate hash tables for the different kinds of entities the DDSI stack deals with (participants, readers and writers for the local entities, and proxy participants, proxy readers and proxy writers for the remote entities), and so it used to be that ``ephash_lookup_participant_guid`` checked a different hash table than ``ephash_lookup_proxy_participant_guid``. At some point I replaced the old implementation of those hash tables by a proper concurrent hash table implementation (an implementation of hopscotch hashing) and unified the 6 tables into a single one. The interface didn't get changed at that time.

Unifying them is possible because each entity has a unique GUID — the U stands for "unique", after all — and so the ability to store different kinds of entities under the same GUID was no longer useful once ``deliver_locally`` was added to deal with communication from a local writer to a local reader. Before that function got added, local communication would actually happen via the network, but that's way slower.

Note that communication via the network may be slower, but is also an clever hack that requires minimal effort to make such local communications work. Often there is but little communication between writers and readers from a single participant, and so it is not a bad hack per se. However, this particular DDSI stack is built to efficiently handle many participants in a single domain, and so is a very nice basis for running DDS-based microservices — and then local communication efficiency does become an issue.
